baseband/gsb/base.py

```diff
@@ -269,82 +269,22 @@ def _last_header(self):
-            ``sample_shape``, are (polarization, channel).  Any
-            dimension of length unity is removed if ``self.squeeze=True``.
-        """
-        if out is None:
-            if count is None or count < 0:
-                count = self.size - self.offset
-                if count < 0:
-                    raise EOFError
-
-            dtype = np.complex64 if self.complex_data else np.float32
-            out = np.empty((count,) + self.sample_shape, dtype)
-        else:
-            assert out.shape[1:] == self.sample_shape, (
-                "'out' should have trailing shape {}".format(self.sample_shape))
-            count = out.shape[0]
-
-        offset0 = self.offset
-        while count > 0:
-            frame_nr, sample_offset = divmod(self.offset,
-                                             self.samples_per_frame)
-            if(frame_nr != self._frame_nr):
-                # Read relevant frame (possibly reusing data array from
-                # previous frame set).
-                self._read_frame()
-                framerate = self.sample_rate / self.samples_per_frame
-                assert np.isclose(self._frame_nr,
-                                  ((self._frame.header.time -
-                                    self.start_time) * framerate).to(u.one))
-
-            # Determine appropriate slice to decode.
-            nsample = min(count, self.samples_per_frame - sample_offset)
-            sample = self.offset - offset0
-            data = self._frame[sample_offset:sample_offset + nsample]
-            data = self._squeeze_and_subset(data)
-            # Copy relevant data from frame into output.
-            out[sample:sample + nsample] = data
-            self.offset += nsample
-            count -= nsample
-
-        return out
-
-    def _read_frame(self, out=None):
-        frame_nr = self.offset // self.samples_per_frame
+    def _read_frame(self, frame_nr):
         self.fh_ts.seek(self.header0.seek_offset(frame_nr))
         if self.header0.mode == 'rawdump':
             self.fh_raw.seek(frame_nr * self._payloadsize)
         else:
             for fh_pair in self.fh_raw:
                 for fh in fh_pair:
                     fh.seek(frame_nr * self._payloadsize)
-        self._frame = GSBFrame.fromfile(self.fh_ts, self.fh_raw,
-                                        payloadsize=self._payloadsize,
-                                        nchan=self._unsliced_shape.nchan,
-                                        bps=self.bps,
-                                        complex_data=self.complex_data)
-        self._frame_nr = frame_nr
-        return self._frame
+        frame = GSBFrame.fromfile(self.fh_ts, self.fh_raw,
+                                  payloadsize=self._payloadsize,
+                                  nchan=self._unsliced_shape.nchan,
+                                  bps=self.bps, complex_data=self.complex_data)
+        assert int(round(((frame.header.time - self.start_time) *
```

Comment:
This may be super-slow compared to VDIF's literal timestamp comparison, but happily it happens much less often for GSB!

Reply:
Yes, that's why I left it...